File: public/packages/dotnet/auth/outbound/WorkloadCertificate/WorkloadLeafClient.cs
// -----------------------------------------------------------------------
// <copyright file="WorkloadLeafClient.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Auth.Outbound.WorkloadCertificate;

using System.Net.Security;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using DcsvIo.D2.Auth.Outbound.Telemetry;
using DcsvIo.D2.Resilience.CircuitBreaker;
using DcsvIo.D2.Resilience.Singleflight;
using DcsvIo.D2.Result;
using DcsvIo.D2.Utilities.Diagnostics;
using JetBrains.Annotations;
using Microsoft.Extensions.Logging;
using NodaTime;

/// <summary>
/// Refresh-ahead implementation of <see cref="IWorkloadLeafSource"/>. Generates a
/// FRESH ECDSA P-256 keypair per reissue, builds a PKCS#10 certificate-signing
/// request, obtains a signed leaf through the host-supplied
/// <see cref="IWorkloadCertificateIssuer"/>, pairs the returned certificate with
/// the LOCAL private key, builds the leaf + issuing intermediate into a presentable
/// chain context, caches it in-memory until it nears expiry, and reissues via the
/// background <see cref="WorkloadLeafRefreshHostedService"/>. Concurrent callers
/// (on-demand + the refresh service) dedup to a single reissue via
/// <c>Singleflight</c>; a <c>CircuitBreaker</c> fast-fails after repeated
/// issuer-unreachable failures; a still-valid cached leaf is served stale while a
/// reissue is attempted.
/// </summary>
/// <remarks>
/// <para>
/// <b>The private key never crosses the issuer seam.</b> The keypair is generated
/// inside this client (fresh per reissue — per-rotation key freshness), the CSR
/// carries only public material, and the issuer returns only certificates. The
/// CSR subject is a fixed placeholder: the issuer structurally ignores it (the
/// leaf's subject-alternative-name is the issuer's authenticated view of this
/// workload), so the client deliberately carries NO identity configuration.
/// </para>
/// <para>
/// <b>Mismatch defense.</b> A returned leaf whose public key does not match the
/// local keypair can never be presented (there is no private key for it), so it
/// is rejected before any cache write — the still-valid cached leaf keeps serving
/// and the reissue counts as a transient failure.
/// </para>
/// <para>
/// The refresh-ahead loop keeps <see cref="WorkloadLeafCache"/> holding a current
/// chain. The gRPC presentation path (<c>AddD2WorkloadCertificate</c>) reads that
/// chain context at CHANNEL BUILD, not per-connection — so a consumer holding a
/// long-lived channel adopts a rotated leaf only by rebuilding the channel.
/// Rebuilding a long-lived channel on rotation is the consumer's responsibility.
/// </para>
/// </remarks>
[MustDisposeResource(false)]
internal sealed class WorkloadLeafClient : IWorkloadLeafSource, IDisposable
{
    // Singleflight key — there is exactly one "reissue this workload's leaf"
    // operation per process, so the key is a constant. Multiple concurrent callers
    // (on-demand + the refresh hosted service) all dedup to one reissue.
    private const string _SINGLEFLIGHT_KEY = "workload-leaf";

    // Fixed CSR subject placeholder. The issuer structurally ignores the CSR's
    // subject (the SAN authority is its authenticated peer view), so the client
    // cannot and need not name itself — no identity knob exists by design.
    private const string _CSR_SUBJECT = "CN=d2-workload";

    private readonly IWorkloadCertificateIssuer r_issuer;
    private readonly WorkloadLeafCache r_cache;
    private readonly ILogger<WorkloadLeafClient> r_logger;
    private readonly TimeProvider r_clock;
    private readonly Singleflight<string, ReissueResult> r_singleflight = new();
    private readonly CircuitBreaker<ReissueResult> r_circuitBreaker;
    private bool _disposed;

    /// <summary>
    /// Initializes a new instance of the <see cref="WorkloadLeafClient"/> class.
    /// </summary>
    /// <param name="issuer">The host-supplied leaf issuer (in-process / harness adapter).</param>
    /// <param name="cache">The shared per-process live-leaf cache.</param>
    /// <param name="logger">The logger.</param>
    /// <param name="clock">The time provider (overridable for tests).</param>
    [MustDisposeResource(false)]
    public WorkloadLeafClient(
        IWorkloadCertificateIssuer issuer,
        WorkloadLeafCache cache,
        ILogger<WorkloadLeafClient> logger,
        TimeProvider clock)
    {
        ArgumentNullException.ThrowIfNull(issuer);
        ArgumentNullException.ThrowIfNull(cache);
        ArgumentNullException.ThrowIfNull(logger);
        ArgumentNullException.ThrowIfNull(clock);

        r_issuer = issuer;
        r_cache = cache;
        r_logger = logger;
        r_clock = clock;

        // 5 consecutive transient-failure ReissueResults → 30 s open (the shared
        // AuthOutboundResilienceDefaults thresholds). Value-based predicate required
        // because ReissueAsync catches exceptions internally and returns
        // ReissueResult.TransientFailure().
        r_circuitBreaker = new CircuitBreaker<ReissueResult>(
            isFailure: static r => !r.Success,
            options: new CircuitBreakerOptions(
                failureThreshold: AuthOutboundResilienceDefaults.FAILURE_THRESHOLD,
                cooldownDuration: AuthOutboundResilienceDefaults.SR_CooldownDuration,
                nowFunc: () => clock.GetUtcNow().ToUnixTimeMilliseconds()));
    }

    /// <inheritdoc/>
    public async ValueTask<D2Result<X509Certificate2>> GetCurrentLeafAsync(
        CancellationToken ct = default)
    {
        ObjectDisposedException.ThrowIf(Volatile.Read(ref _disposed), this);

        var now = Instant.FromDateTimeOffset(r_clock.GetUtcNow());
        var cached = r_cache.TryGet(now);

        if (cached is not null)
        {
            r_logger.WorkloadLeafCacheHit(cached.NotAfter.ToDateTimeOffset());
            return D2Result<X509Certificate2>.Ok(cached.Leaf);
        }

        // Cache miss / expired → reissue via singleflight (outer) wrapping circuit
        // breaker (inner). Singleflight deduplicates concurrent callers; the breaker
        // fast-fails when the issuer has been repeatedly unreachable.
        try
        {
            var reissueResult = await r_singleflight.ExecuteAsync(
                _SINGLEFLIGHT_KEY,
                innerCt => r_circuitBreaker.ExecuteAsync(
                    breakerCt => ReissueAsync(force: false, breakerCt), ct: innerCt),
                ct);

            if (!reissueResult.Success)
                return D2Result<X509Certificate2>.ServiceUnavailable();

            // A sibling may have published a fresher snapshot; read through the
            // cache so all callers converge on the one live handle.
            var current = r_cache.TryGet(Instant.FromDateTimeOffset(r_clock.GetUtcNow()));

            if (current is not null)
                return D2Result<X509Certificate2>.Ok(current.Leaf);

            return D2Result<X509Certificate2>.ServiceUnavailable();
        }
        catch (CircuitOpenException)
        {
            return D2Result<X509Certificate2>.ServiceUnavailable();
        }
    }

    /// <summary>
    /// Forces a reissue of the cached leaf. Called by
    /// <see cref="WorkloadLeafRefreshHostedService"/> on the proactive schedule.
    /// Goes through the same singleflight as on-demand callers, but — unlike the
    /// opportunistic on-demand path — mints a FRESH leaf even when a still-valid
    /// leaf is cached: this is the refresh-ahead entry point, and it fires precisely
    /// while the current leaf is valid but inside the lead-time window, so it MUST
    /// reissue rather than short-circuit on the still-valid cache.
    /// </summary>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>A <see cref="D2Result"/> describing the reissue outcome.</returns>
    public async ValueTask<D2Result> ForceReissueAsync(CancellationToken ct = default)
    {
        ObjectDisposedException.ThrowIf(Volatile.Read(ref _disposed), this);

        try
        {
            var reissueResult = await r_singleflight.ExecuteAsync(
                _SINGLEFLIGHT_KEY,
                innerCt => r_circuitBreaker.ExecuteAsync(
                    breakerCt => ReissueAsync(force: true, breakerCt), ct: innerCt),
                ct);

            return reissueResult.Success
                ? D2Result.Ok()
                : D2Result.ServiceUnavailable();
        }
        catch (CircuitOpenException)
        {
            return D2Result.ServiceUnavailable();
        }
    }

    /// <inheritdoc/>
    public void Dispose()
    {
        Volatile.Write(ref _disposed, true);
    }

    /// <summary>
    /// Builds a live, private-key-bearing leaf <see cref="X509Certificate2"/> by
    /// pairing the issuer-returned certificate with the LOCALLY-generated keypair.
    /// The key was never received and never transmitted — there is no received
    /// secret buffer to import or zero.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <c>CopyWithPrivateKey</c> over an in-memory <see cref="ECDsa"/> yields an
    /// EPHEMERAL-key certificate. Linux/OpenSSL — the deployment target — presents an
    /// ephemeral-key leaf for TLS client authentication without issue, so that path is
    /// left exactly as-is.
    /// </para>
    /// <para>
    /// Windows Schannel, however, refuses an ephemeral-key certificate for client
    /// authentication (the handshake fails with <c>0x8009030E</c> — "No credentials
    /// are available in the security package"). The Windows-only branch therefore
    /// round-trips the leaf through PKCS#12 and re-imports WITHOUT
    /// <see cref="X509KeyStorageFlags.EphemeralKeySet"/> (the very flag Schannel
    /// rejects) and WITHOUT <see cref="X509KeyStorageFlags.PersistKeySet"/>, so the
    /// key lands in a Schannel-usable perishable key container that is deleted when
    /// the returned <see cref="X509Certificate2"/> is disposed — the cache disposes
    /// the superseded / current leaf, so no key container leaks across a refresh.
    /// The PKCS#12 buffer (which carries the private key) is zeroed.
    /// </para>
    /// </remarks>
    /// <param name="material">The issuer-returned certificate material (all public).</param>
    /// <param name="localKey">The locally-generated keypair the leaf certifies.</param>
    /// <returns>The live leaf certificate.</returns>
    private static X509Certificate2 BuildLiveLeaf(WorkloadLeafMaterial material, ECDsa localKey)
    {
        using var certOnly = X509CertificateLoader.LoadCertificate(material.CertificateDer);

        if (!OperatingSystem.IsWindows())
            return certOnly.CopyWithPrivateKey(localKey);

        // Windows-only: re-home the ephemeral key into a Schannel-usable perishable
        // key container via a PKCS#12 round-trip. The intermediate ephemeral cert is
        // disposed and the PKCS#12 buffer (which carries the private key) is zeroed.
        using var ephemeralLeaf = certOnly.CopyWithPrivateKey(localKey);
        var pfx = ephemeralLeaf.Export(X509ContentType.Pkcs12);

        try
        {
            // Exportable == no EphemeralKeySet (Schannel-usable) AND no PersistKeySet
            // (the temporary key container is removed when the cert is disposed).
            return X509CertificateLoader.LoadPkcs12(
                pfx,
                password: null,
                keyStorageFlags: X509KeyStorageFlags.Exportable);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(pfx);
        }
    }

    /// <summary>
    /// Builds the cached snapshot: the live leaf (<see cref="BuildLiveLeaf"/> —
    /// paired with the local keypair), the public issuing intermediate, and — where
    /// the platform supports it — the pre-built
    /// <see cref="SslStreamCertificateContext"/> carrying the full
    /// <c>leaf → intermediate</c> chain. Presenting the chain (not the bare leaf)
    /// is what lets a strict peer's root-anchored chain rebuild complete without a
    /// machine-store-resident intermediate or a network (AIA) fetch.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The intermediate is public (no private key), so it carries no key-container
    /// concern; the leaf carries the locally-generated secret key. The context holds
    /// references to both certs — they MUST stay alive while the context is
    /// presentable, so the snapshot owns them and the cache disposes both only on
    /// swap/dispose.
    /// </para>
    /// <para>
    /// <c>offline: true</c> keeps the chain build store-only — the client never reaches
    /// out to the network for missing issuer certificates. On Linux/OpenSSL (the
    /// deployment target) the context is built from the provided leaf + intermediate
    /// and the full chain is presented on the handshake. On Windows the chain is built
    /// by Schannel outside the process, which refuses to construct a chain context for
    /// a leaf whose internal-CA root is not installed in the OS trust store; that build
    /// throws a <see cref="System.Security.Cryptography.CryptographicException"/>, which
    /// is tolerated here (a null context → bare-leaf presentation) because Windows
    /// cannot transmit an application-supplied intermediate regardless. A Windows host
    /// that needs the chain transmitted installs the CA into the OS store (operator
    /// action), after which this same build succeeds and the chain is presented.
    /// </para>
    /// </remarks>
    /// <param name="material">The issuer-returned certificate material (all public).</param>
    /// <param name="localKey">The locally-generated keypair the leaf certifies.</param>
    /// <returns>The snapshot to publish into the cache.</returns>
    private static WorkloadLeafSnapshot BuildSnapshot(
        WorkloadLeafMaterial material, ECDsa localKey)
    {
        var leaf = BuildLiveLeaf(material, localKey);

        X509Certificate2? intermediate = null;

        try
        {
            intermediate = X509CertificateLoader.LoadCertificate(material.IssuerCertificateDer);

            var chainContext = TryBuildChainContext(leaf, intermediate);

            return new WorkloadLeafSnapshot(leaf, intermediate, chainContext, material.NotAfter);
        }
        catch
        {
            // A malformed intermediate DER (LoadCertificate) — or a non-Windows chain-build
            // failure — throws AFTER the live leaf exists; the leaf carries the secret key,
            // so an un-guarded throw here leaks its Schannel key-container handle. Dispose
            // both owned handles on the error path, then rethrow so ReissueAsync's serve-
            // stale catch treats it as transient (the cached leaf keeps serving).
            leaf.Dispose();
            intermediate?.Dispose();

            throw;
        }
    }

    /// <summary>
    /// Builds the presentable <c>leaf → intermediate</c> chain context, tolerating the
    /// Windows-Schannel case where the context cannot be constructed for an
    /// uninstalled-internal-CA-root leaf.
    /// </summary>
    /// <remarks>
    /// On Linux/OpenSSL this always returns a context (the deployment path). On Windows
    /// it returns null when Schannel rejects the chain build (root not in the OS trust
    /// store) — the only Windows-functional fallback is the bare leaf, since Schannel
    /// will not transmit an application-supplied intermediate without store residency.
    /// </remarks>
    /// <param name="leaf">The live private-key-bearing leaf.</param>
    /// <param name="intermediate">The public issuing intermediate.</param>
    /// <returns>The chain context, or null on the Windows store-dependent path.</returns>
    private static SslStreamCertificateContext? TryBuildChainContext(
        X509Certificate2 leaf, X509Certificate2 intermediate)
    {
        try
        {
            return SslStreamCertificateContext.Create(leaf, [intermediate], offline: true);
        }
        catch (CryptographicException) when (OperatingSystem.IsWindows())
        {
            // Windows-only: Schannel cannot build a chain context for a leaf whose
            // internal-CA root is not in the OS trust store, and cannot transmit an
            // application-supplied intermediate regardless. Fall back to presenting the
            // bare leaf (the prior Windows-functional behavior); Linux is unaffected.
            return null;
        }
    }

    /// <summary>
    /// Returns whether the returned leaf certifies the LOCAL keypair — its
    /// SubjectPublicKeyInfo must equal the local key's, byte for byte.
    /// </summary>
    /// <param name="certificateDer">The issuer-returned leaf certificate DER.</param>
    /// <param name="localKey">The locally-generated keypair.</param>
    /// <returns><see langword="true"/> when the leaf's public key matches the local key.</returns>
    private static bool LeafMatchesLocalKey(byte[] certificateDer, ECDsa localKey)
    {
        using var leaf = X509CertificateLoader.LoadCertificate(certificateDer);
        var leafSpki = leaf.PublicKey.ExportSubjectPublicKeyInfo();
        var localSpki = localKey.ExportSubjectPublicKeyInfo();

        return leafSpki.AsSpan().SequenceEqual(localSpki);
    }

    private async ValueTask<ReissueResult> ReissueAsync(bool force, CancellationToken ct)
    {
        // Opportunistic-path cache re-check (force == false ONLY). A sibling on-demand
        // caller may have populated the cache between the GetCurrentLeafAsync TryGet and
        // this Singleflight body; suppressing a redundant mint there is correct — an
        // on-demand caller only needs SOME non-expired leaf, and one now exists.
        //
        // The refresh-ahead FORCE path deliberately SKIPS this suppression. It fires
        // precisely when the cached leaf is still valid but inside the lead-time window,
        // so a "still-valid cached leaf" is exactly the state in which it MUST proceed to
        // mint a fresh leaf ahead of expiry; short-circuiting here would defeat
        // refresh-ahead entirely (the leaf would only ever reissue at/after expiry,
        // stalling on-demand callers on a synchronous mint on the hot path).
        //
        // Concurrency: both paths share one Singleflight key, so a force + on-demand race
        // dedups to a single mint (no double-mint; the cache's Interlocked swap makes the
        // publish torn-read-free). A valid-cache on-demand caller never enters this body
        // (it returns the cached leaf at the outer GetCurrentLeafAsync TryGet), so a force
        // follower can never be silently suppressed by a non-force Singleflight leader.
        if (!force)
        {
            var preReissueCache = r_cache.TryGet(Instant.FromDateTimeOffset(r_clock.GetUtcNow()));

            if (preReissueCache is not null)
                return ReissueResult.Successful();
        }

        try
        {
            // 1) Fresh keypair per reissue — the workload owns its key lifecycle;
            //    rotation freshness holds because a new key is minted every cycle.
            using var localKey = ECDsa.Create(ECCurve.NamedCurves.nistP256);

            // 2) Build the PKCS#10 CSR (public key + placeholder subject +
            //    self-signature — public material by construction).
            var csrRequest = new CertificateRequest(
                _CSR_SUBJECT, localKey, HashAlgorithmName.SHA256);

            var csrDer = csrRequest.CreateSigningRequest();

            // 3) Obtain the signed leaf. Only the CSR crosses the seam.
            var issuance = await r_issuer.IssueAsync(csrDer, ct);

            if (!issuance.Success || issuance.Data is null)
                return ReissueResult.TransientFailure();

            // 4) Mismatch defense: a leaf certifying a DIFFERENT key than the local
            //    one can never be presented (no private key exists for it) — reject
            //    BEFORE any cache write; the still-valid cached leaf keeps serving.
            if (!LeafMatchesLocalKey(issuance.Data.CertificateDer, localKey))
            {
                r_logger.WorkloadLeafIssuerKeyMismatch();
                OutboundTelemetry.SR_LeafReissueFailures.Add(1);
                return ReissueResult.TransientFailure();
            }

            // 5) Pair the returned certificate with the local key and publish. The
            //    local handle is disposed by the using once the live cert owns the
            //    key (CopyWithPrivateKey duplicates it into the certificate).
            r_cache.Set(BuildSnapshot(issuance.Data, localKey));

            return ReissueResult.Successful();
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception ex)
        {
            // Serve-stale contract: ANY exception generating the keypair, building
            // the CSR, or building the snapshot — the live certificate
            // (CryptographicException, ArgumentException, InvalidOperationException
            // from CopyWithPrivateKey algorithm mismatch), decoding the intermediate
            // DER, or constructing the chain context (NotSupportedException if the
            // leaf lacks a private key) — is treated as transient: the cached leaf
            // keeps being served while the next reissue cycle may succeed.
            // OperationCanceledException propagates above.
            // Sanitize: never log ex itself (a cert-parse exception could echo
            // subject / SAN content). Type FullName + first frame are safe.
            // The cached leaf's not-after is captured as a structured log field
            // (ISO-8601 UTC) so operators can correlate failures with expiry proximity.
            // It cannot be a metric tag (timestamps are high-cardinality, not enumerable
            // dimensions); the log record is the correct OTel home for this value.
            var staleCached = r_cache.PeekRaw();

            var cachedLeafNotAfter = staleCached is not null
                ? staleCached.NotAfter.ToDateTimeOffset().ToString("O")
                : "none";

            r_logger.WorkloadLeafReissueFailed(
                SanitizedExceptionRender.TypeName(ex),
                SanitizedExceptionRender.FirstFrame(ex),
                cachedLeafNotAfter);

            OutboundTelemetry.SR_LeafReissueFailures.Add(1);

            return ReissueResult.TransientFailure();
        }
    }

    /// <summary>
    /// Reissue outcome. <c>internal</c> (not <c>private</c>) so the assembly-level
    /// <see cref="CircuitBreaker{T}"/> generic can be instantiated with this type
    /// from within the same assembly.
    /// </summary>
    internal readonly record struct ReissueResult(bool Success)
    {
        /// <summary>Returns a successful reissue outcome.</summary>
        public static ReissueResult Successful() => new(true);

        /// <summary>Returns a transient-failure reissue outcome.</summary>
        public static ReissueResult TransientFailure() => new(false);
    }
}

File: private/services/edge/key-custodian/client-ts/src/issuance/workload-leaf-client.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import {
  type D2Result,
  ok,
  canceled,
  serviceUnavailable,
  bubbleFail,
} from "@dcsv-io/d2-result";
import { CircuitBreaker, Singleflight } from "@dcsv-io/d2-resilience";
import { sanitizedErrorRender, type ILogger } from "@dcsv-io/d2-logging";
import type { ChannelCredentials } from "@grpc/grpc-js";
import type { WorkloadCertificateIssuer } from "./workload-certificate-issuer.js";
import type { CaTrustBundle, LeafSnapshot } from "./workload-leaf-material.js";
import type { WorkloadLeafClientOptions } from "./leaf-client-options.js";
import { WorkloadLeafCache } from "./leaf-cache.js";
import { generateLeafKeypair } from "./leaf-keypair.js";
import { buildCsr, MAX_CSR_DER_BYTES } from "./csr-builder.js";
import { leafMatchesLocalKey } from "./leaf-key-match.js";
import { derToPem } from "./der-pem.js";
import { assembleTrustStore } from "./trust-assembly.js";
import { buildMutualTlsCredentials } from "./mtls-channel.js";

// Exactly one "reissue this workload's leaf" operation per process — the
// singleflight key is a constant so on-demand + refresh-ahead callers dedup.
const _SINGLEFLIGHT_KEY = "workload-leaf";

// Refresh-ahead margin + circuit defaults, mirroring the .NET twins verbatim.
// Provenance (value): AuthOutboundOptions.WorkloadLeafRefreshLeadTime = 5 min;
// AuthOutboundResilienceDefaults.FAILURE_THRESHOLD = 5; .SR_CooldownDuration = 30 s.
// Dual drift-guard: workload-leaf-defaults.test.ts pins these literals here and the
// .NET AuthOutboundDefaultsParityTests pins the twin - divergence reds a pin test.
export const DEFAULT_REFRESH_MARGIN_MS = 5 * 60 * 1000;
export const DEFAULT_FAILURE_THRESHOLD = 5;
export const DEFAULT_COOLDOWN_MS = 30 * 1000;

/**
 * Refresh-ahead workload-leaf client — the TS twin of the .NET
 * `WorkloadLeafClient`. Generates a FRESH ECDSA P-256 keypair per reissue (the
 * private key never leaves the process), builds a PKCS#10 CSR, obtains a signed
 * leaf through the injected {@link WorkloadCertificateIssuer} port, verifies the
 * returned leaf certifies the local key (mismatch defense), caches the presentable
 * leaf until it nears expiry, and reissues under a single-flight + circuit-breaker
 * with serve-stale-on-transient semantics. Also fetches + caches the CA trust
 * bundle and assembles the mutual-TLS channel credentials the workload presents.
 *
 * The private key never crosses the issuer seam — only the CSR (public material)
 * does. A returned leaf whose public key does not match the local keypair is
 * rejected before any cache write (there is no private key for it), so the
 * still-valid cached leaf keeps serving.
 */
export class WorkloadLeafClient {
  readonly #issuer: WorkloadCertificateIssuer;
  readonly #cache = new WorkloadLeafCache();
  readonly #singleflight = new Singleflight<string, boolean>();
  readonly #circuitBreaker: CircuitBreaker<boolean>;
  readonly #now: () => number;
  readonly #refreshMarginMs: number;
  readonly #logger: ILogger | undefined;
  #caTrustBundle: CaTrustBundle | undefined;
  #disposed = false;

  /**
   * @param issuer  - The transport port that issues leaves + fetches the CA chain.
   * @param options - Tunables (clock, refresh margin, circuit thresholds, logger).
   */
  constructor(
    issuer: WorkloadCertificateIssuer,
    options: WorkloadLeafClientOptions = {},
  ) {
    this.#issuer = issuer;
    this.#now = options.now ?? Date.now;
    this.#refreshMarginMs =
      options.refreshMarginMs ?? DEFAULT_REFRESH_MARGIN_MS;
    this.#logger = options.logger;
    this.#circuitBreaker = new CircuitBreaker<boolean>({
      failureThreshold:
        options.circuitFailureThreshold ?? DEFAULT_FAILURE_THRESHOLD,
      cooldownMs: options.circuitCooldownMs ?? DEFAULT_COOLDOWN_MS,
      isFailure: (r) => !r,
      nowFunc: this.#now,
    });
  }

  /**
   * Return the current presentable leaf, reissuing when the cache is empty /
   * expired or within the refresh-ahead margin. A still-valid cached leaf is
   * served even when the reissue fails (serve-stale). Returns a typed
   * `serviceUnavailable` when no valid leaf can be produced, or `canceled` when
   * the caller's signal aborts.
   */
  async getCurrentLeaf(signal?: AbortSignal): Promise<D2Result<LeafSnapshot>> {
    this.#throwIfDisposed();

    const nowMs = this.#now();
    const cached = this.#cache.tryGet(nowMs);

    if (cached !== undefined && !this.#needsRefresh(cached, nowMs))
      return ok(cached);

    return this.#reissueThenRead(signal);
  }

  /**
   * Force a reissue (the proactive-refresh entry point), going through the same
   * single-flight as on-demand callers. Returns `ok` when a valid leaf is present
   * afterward, else `serviceUnavailable` / `canceled`.
   */
  async forceReissue(signal?: AbortSignal): Promise<D2Result<void>> {
    this.#throwIfDisposed();

    const result = await this.#reissueThenRead(signal);

    return result.success ? ok() : bubbleFail<void>(result);
  }

  /**
   * Fetch + cache the CA trust bundle (root + issuing intermediate) the workload
   * pins when validating the mutual-TLS server. Cached for the client's lifetime
   * — the CA chain is long-lived trust material.
   */
  async getCaTrustBundle(
    signal?: AbortSignal,
  ): Promise<D2Result<CaTrustBundle>> {
    this.#throwIfDisposed();

    if (this.#caTrustBundle !== undefined) return ok(this.#caTrustBundle);

    const chain = await this.#issuer.getCaCertificate(signal);

    if (chain.failed) return bubbleFail<CaTrustBundle>(chain);

    // A success envelope with no data is a malformed response — typed retryable
    // failure, never an empty trust store.
    if (chain.data === undefined) return serviceUnavailable<CaTrustBundle>();

    const bundle = assembleTrustStore(chain.data);
    this.#caTrustBundle = bundle;

    return ok(bundle);
  }

  /**
   * Assemble mutual-TLS channel credentials presenting the current leaf chain +
   * private key and pinning the fetched CA bundle — the workload's client-side
   * mutual-TLS presentation. Reissues / fetches as needed.
   */
  async currentChannelCredentials(
    signal?: AbortSignal,
  ): Promise<D2Result<ChannelCredentials>> {
    const leaf = await this.getCurrentLeaf(signal);

    if (leaf.failed) return bubbleFail<ChannelCredentials>(leaf);

    const trust = await this.getCaTrustBundle(signal);

    if (trust.failed) return bubbleFail<ChannelCredentials>(trust);

    const credentials = buildMutualTlsCredentials({
      caBundlePem: trust.data!.caBundlePem,
      certChainPem: leaf.data!.certChainPem,
      privateKeyPem: leaf.data!.privateKeyPem,
    });

    return ok(credentials);
  }

  /** Marks the client disposed — subsequent calls throw. */
  dispose(): void {
    this.#disposed = true;
  }

  // -------------------------------------------------------------------------
  // Internals
  // -------------------------------------------------------------------------

  async #reissueThenRead(
    signal?: AbortSignal,
  ): Promise<D2Result<LeafSnapshot>> {
    try {
      await this.#singleflight.do(_SINGLEFLIGHT_KEY, () =>
        this.#circuitBreaker.execute(() => this.#reissue(signal)),
      );
    } catch (e) {
      // Abort → surface cancellation. The only other exception the
      // single-flight + breaker pipeline raises is CircuitOpenError (fast-fail
      // while open); swallow it and fall through to the read-through so a
      // still-valid cached leaf is served stale.
      if (this.#isAbort(e, signal)) return canceled<LeafSnapshot>();
    }

    const current = this.#cache.tryGet(this.#now());

    if (current !== undefined) return ok(current);

    return serviceUnavailable<LeafSnapshot>();
  }

  async #reissue(signal?: AbortSignal): Promise<boolean> {
    // A sibling caller may have published a FRESH leaf between the caller's tryGet
    // and the single-flight entry — short-circuit only for a leaf that does NOT
    // itself need refreshing (a still-within-margin leaf must proceed to reissue,
    // else the refresh-ahead cycle would never fire).
    const preReissue = this.#cache.tryGet(this.#now());

    if (
      preReissue !== undefined &&
      !this.#needsRefresh(preReissue, this.#now())
    )
      return true;

    try {
      // Fresh keypair per reissue — rotation freshness holds because a new key is
      // minted every cycle; the private key never crosses the issuer seam.
      const keypair = await generateLeafKeypair();
      const csrDer = await buildCsr(keypair.cryptoKeyPair);

      /* v8 ignore start -- a real P-256 CSR is well under the 4 KiB cap; the client-side pre-check is defensive (the server enforces the same cap) */
      if (csrDer.byteLength > MAX_CSR_DER_BYTES) {
        this.#logReissueFailure(
          new Error("csr-exceeds-cap"),
          "csr-exceeds-cap",
        );
        return false;
      }
      /* v8 ignore stop */

      const issuance = await this.#issuer.issueLeaf(csrDer, signal);

      if (issuance.failed || issuance.data === undefined) return false;

      const material = issuance.data;

      // Mismatch defense: a leaf certifying a DIFFERENT key can never be
      // presented (no private key for it) — reject before any cache write.
      if (!leafMatchesLocalKey(material.certificateDer, keypair.spkiDer)) {
        this.#logKeyMismatch();
        return false;
      }

      const privateKeyPem = await keypair.exportPrivateKeyPkcs8Pem();
      const certChainPem =
        derToPem(material.certificateDer, "CERTIFICATE") +
        derToPem(material.issuerCertificateDer, "CERTIFICATE");

      this.#cache.set({
        certChainPem,
        privateKeyPem,
        notAfter: material.notAfter,
      });

      return true;
    } catch (e) {
      // Abort propagates (the outer pipeline maps it to cancellation). Any other
      // error — keygen / CSR / issuer / cert parse / export — is transient: the
      // cached leaf keeps serving while the next reissue may succeed.
      if (this.#isAbort(e, signal)) throw e;

      this.#logReissueFailure(e, undefined);
      return false;
    }
  }

  #needsRefresh(cached: LeafSnapshot, nowMs: number): boolean {
    return cached.notAfter.epochMilliseconds - this.#refreshMarginMs <= nowMs;
  }

  #isAbort(_e: unknown, signal: AbortSignal | undefined): boolean {
    return signal?.aborted === true;
  }

  #throwIfDisposed(): void {
    if (this.#disposed) throw new Error("WorkloadLeafClient has been disposed");
  }

  #logKeyMismatch(): void {
    // Never logs key material — a fixed operator-meaningful message only.
    this.#logger?.warn(
      "Workload leaf reissue rejected: issuer returned a leaf certifying a different key",
    );
  }

  #logReissueFailure(error: unknown, reason: string | undefined): void {
    const stale = this.#cache.peekRaw();
    const cachedLeafNotAfter =
      stale !== undefined ? stale.notAfter.toString() : "none";

    // Sanitized error shape only — never the raw error (a cert-parse error could
    // echo subject / SAN content); never key material.
    const render = sanitizedErrorRender(error);

    this.#logger?.error("Workload leaf reissue failed", {
      cachedLeafNotAfter,
      reason,
      errorName: render.name,
      errorFrame: render.firstFrame,
    });
  }
}
